api/src/license_key.rs

```diff
@@ -0,0 +1,108 @@
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum LicenseKeyStatus {
+    Valid,
+    Invalid,
+    Blacklisted,
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct LicenseKeyProperties {
+    pub key_size: usize,
+    pub payload_size: usize,
+    pub checksum_size: usize,
+}
+
+impl LicenseKeyProperties {
+    pub fn default() -> LicenseKeyProperties {
+        LicenseKeyProperties {
+            key_size: 0,
+            payload_size: 0,
+            checksum_size: 0,
+        }
+    }
+}
+
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct LicenseKey {
+    pub key: Vec<u8>,
+    pub payload: Vec<u8>,
+    pub checksum: Vec<u8>,
+    pub properties: LicenseKeyProperties,
+    pub serialized_key: Vec<u8>,
+}
+
+impl LicenseKey {
+    #[inline(always)]
+    pub fn deserialize(
+        raw_key: &Vec<u8>,
+        key_size: usize,
+        payload_size: usize,
+        checksum_size: usize,
+    ) -> Result<LicenseKey, &'static str> {
+        if raw_key.len() < (key_size + payload_size + checksum_size) {
+            return Err(
+                "Cannot deserialize license key with larger properties than raw key itself!",
+            );
+        }
+
+        let properties = LicenseKeyProperties {
+            key_size,
+            payload_size,
+            checksum_size,
```

Comment:
This is configurable? Because in other part of the code you harcoded the value to be 4.

Reply:
Yeah, you are right about that, What I forget to mention is that, the checksum will be user/dev defined function for generating checksum and also user/dev will define what is the size. So it was my unfinished design/code.

But I finished design for it, so it should not be so confusing...

```rust
let license_key_hash_size = self.key_size - self.checksum_byte_size - self.magic.payload_size();
```